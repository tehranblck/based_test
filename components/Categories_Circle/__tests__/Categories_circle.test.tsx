import React from 'react'
import { render, screen, fireEvent } from '@testing-library/react'
import CategoriesCircle, { type CategoryItem } from '@/components/Categories_Circle/Categories_circle'

jest.useFakeTimers()

function getTrack(container: HTMLElement) {
    const track = container.querySelector('[class*="overflow-x-auto"]') as HTMLDivElement | null
    if (!track) throw new Error('track not found')
    return track
}

describe('CategoriesCircle', () => {
    it('renders items and supports links', () => {
        const items: CategoryItem[] = [
            { src: '/slide_2.png', alt: 'cat-1', href: '/c/1' },
            { src: '/slide_2.png', alt: 'cat-2' },
        ]
        const { container } = render(<CategoriesCircle items={items} size={48} />)
        expect(container).toMatchSnapshot()
        expect(screen.getByRole('link', { name: 'cat-1' })).toBeInTheDocument()
    })

    it('auto-scrolls when content overflows and not hovering', () => {
        const { container } = render(<CategoriesCircle size={48} speedMsPerStep={10} />)
        const track = getTrack(container)
        Object.defineProperty(track, 'clientWidth', { value: 100, configurable: true })
        Object.defineProperty(track, 'scrollWidth', { value: 300, configurable: true })
        Object.defineProperty(track, 'scrollLeft', { value: 0, writable: true })

        fireEvent.mouseEnter(track)
        fireEvent.mouseLeave(track)

        jest.advanceTimersByTime(12)
        expect(track.scrollLeft).toBeGreaterThan(0)
    })

    it('manual prev/next buttons call scrollTo with smooth behavior', () => {
        const items: CategoryItem[] = new Array(5).fill(0).map((_, i) => ({ src: '/slide_2.png', alt: `cat-${i}` }))
        const { container } = render(<CategoriesCircle items={items} size={60} />)
        const track = getTrack(container)
        Object.defineProperty(track, 'clientWidth', { value: 120, configurable: true })
        Object.defineProperty(track, 'scrollWidth', { value: 800, configurable: true })
        Object.defineProperty(track, 'scrollLeft', { value: 0, writable: true })
        // @ts-expect-error jsdom
        track.scrollTo = jest.fn()

        fireEvent.click(screen.getByRole('button', { name: 'Next' }))
        expect((track.scrollTo as jest.Mock).mock.calls[0][0]).toMatchObject({ left: expect.any(Number), behavior: 'smooth' })

        fireEvent.click(screen.getByRole('button', { name: 'Prev' }))
        expect((track.scrollTo as jest.Mock).mock.calls[1][0]).toMatchObject({ left: expect.any(Number), behavior: 'smooth' })
    })
})



