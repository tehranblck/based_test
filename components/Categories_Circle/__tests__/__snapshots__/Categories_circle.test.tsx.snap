// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`CategoriesCircle renders items and supports links 1`] = `
<div>
  <section
    class="w-full py-6"
  >
    <div
      class="relative mx-auto"
    >
      <div
        class="relative overflow-hidden"
        style="mask-image: linear-gradient(90deg, transparent, black 6%, black 94%, transparent);"
      >
        <div
          class="flex flex-nowrap gap-3 overflow-x-auto scroll-smooth snap-x snap-mandatory [-ms-overflow-style:none] [scrollbar-width:none] [&::-webkit-scrollbar]:hidden"
        >
          <a
            aria-label="cat-1"
            class="outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:rounded-2xl"
            href="/c/1"
          >
            <div
              class="group relative shrink-0 snap-start rounded-2xl p-[1px] bg-gradient-to-br from-primary/40 via-transparent to-primary/40"
              style="width: 48px; height: 48px;"
            >
              <div
                class="relative h-full w-full rounded-[14px] bg-card/70 backdrop-blur-sm shadow-sm transition-all duration-200 group-hover:shadow-lg group-hover:-translate-y-0.5"
              >
                <div
                  class="absolute inset-0 rounded-[14px] opacity-0 group-hover:opacity-100 transition-opacity duration-200 bg-gradient-to-br from-primary/10 via-transparent to-primary/20"
                />
                <img
                  alt="cat-1"
                  class="object-cover rounded-[14px] transition-transform duration-300 group-hover:scale-[1.06]"
                  data-nimg="fill"
                  decoding="async"
                  loading="lazy"
                  sizes="48px"
                  src="/_next/image?url=%2Fslide_2.png&w=3840&q=75"
                  srcset="/_next/image?url=%2Fslide_2.png&w=16&q=75 16w, /_next/image?url=%2Fslide_2.png&w=32&q=75 32w, /_next/image?url=%2Fslide_2.png&w=48&q=75 48w, /_next/image?url=%2Fslide_2.png&w=64&q=75 64w, /_next/image?url=%2Fslide_2.png&w=96&q=75 96w, /_next/image?url=%2Fslide_2.png&w=128&q=75 128w, /_next/image?url=%2Fslide_2.png&w=256&q=75 256w, /_next/image?url=%2Fslide_2.png&w=384&q=75 384w, /_next/image?url=%2Fslide_2.png&w=640&q=75 640w, /_next/image?url=%2Fslide_2.png&w=750&q=75 750w, /_next/image?url=%2Fslide_2.png&w=828&q=75 828w, /_next/image?url=%2Fslide_2.png&w=1080&q=75 1080w, /_next/image?url=%2Fslide_2.png&w=1200&q=75 1200w, /_next/image?url=%2Fslide_2.png&w=1920&q=75 1920w, /_next/image?url=%2Fslide_2.png&w=2048&q=75 2048w, /_next/image?url=%2Fslide_2.png&w=3840&q=75 3840w"
                  style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
                />
              </div>
            </div>
          </a>
          <div
            class="group relative shrink-0 snap-start rounded-2xl p-[1px] bg-gradient-to-br from-primary/40 via-transparent to-primary/40"
            style="width: 48px; height: 48px;"
          >
            <div
              class="relative h-full w-full rounded-[14px] bg-card/70 backdrop-blur-sm shadow-sm transition-all duration-200 group-hover:shadow-lg group-hover:-translate-y-0.5"
            >
              <div
                class="absolute inset-0 rounded-[14px] opacity-0 group-hover:opacity-100 transition-opacity duration-200 bg-gradient-to-br from-primary/10 via-transparent to-primary/20"
              />
              <img
                alt="cat-2"
                class="object-cover rounded-[14px] transition-transform duration-300 group-hover:scale-[1.06]"
                data-nimg="fill"
                decoding="async"
                loading="lazy"
                sizes="48px"
                src="/_next/image?url=%2Fslide_2.png&w=3840&q=75"
                srcset="/_next/image?url=%2Fslide_2.png&w=16&q=75 16w, /_next/image?url=%2Fslide_2.png&w=32&q=75 32w, /_next/image?url=%2Fslide_2.png&w=48&q=75 48w, /_next/image?url=%2Fslide_2.png&w=64&q=75 64w, /_next/image?url=%2Fslide_2.png&w=96&q=75 96w, /_next/image?url=%2Fslide_2.png&w=128&q=75 128w, /_next/image?url=%2Fslide_2.png&w=256&q=75 256w, /_next/image?url=%2Fslide_2.png&w=384&q=75 384w, /_next/image?url=%2Fslide_2.png&w=640&q=75 640w, /_next/image?url=%2Fslide_2.png&w=750&q=75 750w, /_next/image?url=%2Fslide_2.png&w=828&q=75 828w, /_next/image?url=%2Fslide_2.png&w=1080&q=75 1080w, /_next/image?url=%2Fslide_2.png&w=1200&q=75 1200w, /_next/image?url=%2Fslide_2.png&w=1920&q=75 1920w, /_next/image?url=%2Fslide_2.png&w=2048&q=75 2048w, /_next/image?url=%2Fslide_2.png&w=3840&q=75 3840w"
                style="position: absolute; height: 100%; width: 100%; left: 0px; top: 0px; right: 0px; bottom: 0px; color: transparent;"
              />
            </div>
          </div>
        </div>
        <div
          class="absolute z-50 inset-y-0 left-0 right-0 hidden items-center justify-between sm:flex"
        >
          <button
            aria-label="Prev"
            class="h-8 z-50 w-8 -translate-x-2 rounded-full bg-card/70 text-foreground shadow-sm ring-1 ring-border backdrop-blur hover:bg-card focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring"
            type="button"
          >
            ‹
          </button>
          <button
            aria-label="Next"
            class="h-8 z-50 w-8 translate-x-0 rounded-full bg-card/70 text-foreground shadow-sm ring-1 ring-border backdrop-blur hover:bg-card focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring"
            type="button"
          >
            ›
          </button>
        </div>
      </div>
    </div>
  </section>
</div>
`;
